### PR TITLE
diag(p3): hidden_size sweep driver + analyzer for BC-fit capacity probe (#280)

### LINKED ISSUE #280: diag(p3): hidden_size sweep for BC-fit (Phase 1.5 escalation if #278 fails) — test capacity hypothesis
## Hypothesis being tested

If class-balanced BC (#279) still fails the WORK-house discriminating sub-task, then the failure is **capacity-bound**, not loss-distribution-bound. This experiment tests whether scaling `hidden_size` (current default 64, ~8.1k params) fixes the representational gap.

## Context (Curator update — 2026-05-17)

PR #278's BC-init builder produced a **cross-finding** that may obviate the original 4-cell sweep:

- 10k pairs / 10 epochs @ `hidden_size=64` → `gap_closed = 0.368`, house-on-WORK = 31.5% — reproduces #272's failure mode
- 40k pairs / 30 epochs @ `hidden_size=64` → `gap_closed = 0.934` — **BC succeeded at the default size**

This suggests the architecture has sufficient capacity at default `hidden_size=64`; the original "INDUCTIVE_BIAS_GAP" verdict from #272 (PR #276) is more likely a training-budget artifact than a representational ceiling. Builder has discretion to scope this experiment down (see "Recommended scope" below).

## Recommended scope (after #278 cross-finding)

Two paths, in increasing cost order. **Pick the smallest path that resolves the open question for the cell where #279 (class-balanced BC) still falls short.**

### Path A (minimal — recommended default): 2-cell capacity probe at short-budget regime

Goal: confirm capacity is not the bottleneck when budget is short. Compare at the low-budget setting where the failure originally appeared.

- `hidden_size ∈ {64, 128}` at **10k pairs / 10 epochs** (the budget where #272/PR #276 saw failure)
- If 128 lifts house-on-WORK significantly while 64 stays at chance → capacity matters at short budget (file: increase default and/or use longer budget in production)
- If both stay at chance → capacity isn't the limiting factor at short budget; class imbalance or data quantity dominates → defer to #279/#281 findings

### Path B (full sweep — only if Path A is inconclusive or #279 verdict 3 demands it)

- `hidden_size ∈ {64, 128, 256, 512}` at 10 epochs (extend to 30 if loss still decreasing — PR #276 noted loss was actively decreasing at epoch 10, *underfitting* not overfitting)
- Use class-balanced loss if #279's verdict indicated partial improvement (see "What BC loss to track" below)

For each cell, report: final eval loss, house-on-WORK accuracy, joint accuracy, parameter count.

## What BC loss to track

Use `bc_fit_only.py`'s existing per-head accuracy breakdown (see `compute_verdict`):

- `final.eval_loss` — sum-of-CE over the three heads `[house, mode, signal]`
- `final.house_acc` — house head, all rows (dominated by REST class prior P(REST)≈0.9435)
- `final.mode_acc` — REST/WORK 2-way
- `final.signal_acc` — 2-way
- `final.joint_acc` — exact match across all 3 heads
- **`result.house_acc_on_work_subset`** — the discriminating metric. This is the eval-set accuracy of the house head restricted to rows where the specialist labeled mode=WORK. Chance level for 3 owned houses is ≈0.354. A successful fit must reach ≥0.90 here. **This is the headline number for verdict logic.**
- `result.n_params` — for capacity-vs-accuracy plot

## Verdict logic (capacity-vs-data interpretation)

| house-on-WORK pattern across cells | Interpretation | Next step |
|---|---|---|
| **64 already saturates (≥0.90)** at short budget | Capacity wasn't the gap. If #279 also passed → resolved by class-balancing. If #279 failed → reframe research direction; specialist may use info not in obs. | Close ladder. File ppo-help-brief update. |
| **Monotonic improvement with hidden_size**, plateau ≥0.90 by 128 or 256 | Confirmed capacity bound at short budget. PPO's default-64 actor is undersized for this discriminating sub-task. | File production-PPO follow-up with winning `hidden_size` (likely 128 or 256) |
| **Plateau below 0.90 even at 512** | Architecture mismatch isn't pure capacity — it's *kind* of capacity (depth? attention? non-linearity?). | Promote #281 (TransformerPolicyNetwork). |
| **No cell exceeds chance (~0.35)** | Neither capacity nor class-balance #279 is the gap. Information for the specialist's argmax over burning-owned houses may not be in the observation. | Reframe: inspect what `obs[i]` actually contains; specialist may use privileged info. |

## Implementation guidance

`experiments/p3_specialization/bc_fit_only.py` already accepts `--hidden-size` (default 64) and emits the metrics above. The sweep is a thin bash wrapper:

```bash
for HS in 64 128; do  # or 64 128 256 512 for Path B
  uv run python experiments/p3_specialization/bc_fit_only.py \
    --hidden-size "$HS" \
    --num-steps 2500 --epochs 10 \
    --out "experiments/p3_specialization/runs/280_hs${HS}.json"
done
```

If #279's verdict says class-balancing helped partially: add a `--class-balanced` flag to `bc_fit_only.py` (mirror whatever #279 ships) and pass it for all cells. Otherwise run the existing code as-is.

Aggregate the JSON outputs and report the table in the issue comment / notebook.

## Affected Files

- `experiments/p3_specialization/bc_fit_only.py` — existing; runner accepts `--hidden-size`. No changes expected unless adding `--class-balanced` per #279's outcome (see `bc_fit_only.py:287` for arg, `bc_fit_only.py:136-178` for `train_bc`).
- `bucket_brigade/training/networks.py` — reference only. `PolicyNetwork` definition at lines 13-99; `hidden_size: int = 64` default at line 45.
- `experiments/p3_specialization/runs/280_hs{64,128,...}.json` — new per-cell output files.
- `experiments/p3_specialization/analyze_280.py` (new, optional) — aggregator that reads the per-cell JSONs and emits the verdict per the matrix above. Pattern: see `experiments/p3_specialization/analyze_270.py`.
- Project memory: `~/.claude/projects/-Users-rwalters-GitHub-bucket-brigade/memory/project_ppo_failure_mode.md` — update with verdict outcome.

## Test Plan

- [ ] Manual verification: run Path A (2 cells). Confirm each cell emits a valid result JSON with `result.house_acc_on_work_subset` present and finite.
- [ ] Automated tests: N/A (this is a diagnostic experiment, not production code). If a `--class-balanced` flag is added to `bc_fit_only.py`, add a smoke test alongside whatever #279 ships.
- [ ] Verdict applied: produce a markdown table of `{hidden_size, n_params, eval_loss, house_acc, house_acc_on_work_subset, joint_acc}` for each cell and apply the verdict from the matrix above.
- [ ] Cross-reference: if winning `hidden_size > 64`, file a production-PPO follow-up issue that uses the winning size in `experiments/p3_specialization/train.py`. If `64` saturates and #279 also passed, comment on this issue with the resolution and link to the relevant project-memory update.
- [ ] Edge case: confirm that `hidden_size=512` PolicyNetwork still loads and trains under the existing trainer (no shape mismatch with PPO's downstream consumers if production follow-up is filed).

## Dependencies

- [ ] #279: Class-balanced BC verdict (must complete first — determines whether to use `--class-balanced` flag in this sweep, and whether this experiment is needed at all)

If #279 shows class-balancing already lifts house-on-WORK to ≥0.90 at `hidden_size=64`, this issue can be closed as resolved without running cells.

Note: PR #278's BC-init cross-finding (40k/30ep at hidden_size=64 → gap_closed 0.934) is *advisory context*, not a strict dependency — it suggests but does not prove that #280 is unnecessary at full training budget.

## Cost

- Path A: ~30-60 min total (2 cells × ~15-30 min each, pure CPU).
- Path B: ~2-3 hours total (4 cells × ~30 min each, pure CPU).

Run on `COMPUTE_HOST_PRIMARY` per CLAUDE.md's compute guidelines. Do not run locally.

## References

- PR #276 (the unbalanced #272 baseline this builds on)
- PR #278 (BC-init builder's cross-finding: BC succeeded at default `hidden_size=64` with 40k/30ep)
- #279 (class-balanced BC — must complete first; verdict determines whether to use class-balancing here)
- #281 (TransformerPolicyNetwork escalation — gated on this issue's verdict)
- `bucket_brigade/training/networks.py:13-99` (PolicyNetwork; default `hidden_size=64`)
- `experiments/p3_specialization/bc_fit_only.py:275-298` (CLI arg parsing)
- `docs/ppo-help-brief.md`

### PR DESCRIPTION
**Pull request by rjwalters**
## Summary

Adds the Path A 2-cell sweep driver, verdict-matrix analyzer, and tests for issue #280's BC-fit capacity probe. `bc_fit_only.py` already exposes `--hidden-size` (lines 287, 333), so no changes to the experiment script itself.

Closes #280. Note: #278's 0.934 at hidden_size=64 suggests capacity is not the bottleneck — Path A 2-cell probe will confirm cheaply.

## Changes

- `experiments/p3_specialization/run_issue280_hidden_size_sweep.sh` — Path A default: `HIDDEN_SIZES="64 128"`, `NUM_STEPS=2500`, `EPOCHS=10` (10k pairs × 10 epochs — the short budget where #272/PR #276 first saw failure). Path B accessible via `HIDDEN_SIZES="64 128 256 512"` env override. Modeled on `run_issue262_sweep.sh`: xargs `-P` parallelism, per-cell `train.log`, git-rev stamping into result JSON.
- `experiments/p3_specialization/analyze_280.py` — Aggregator that reads `runs/issue280_hidden_size/hs_{HS}/seed_{SEED}/result.json` and emits `analysis.json` + `verdict.md`. Verdict matrix tied to `house_acc_on_work_subset` (chance ≈ 0.354, success ≥ 0.90), classifies one of: `capacity_not_gap`, `capacity_bound`, `architecture_mismatch`, `information_gap`, `partial`, `no_data`.
- `tests/test_bc_fit_only_hidden_size.py` — 6 tests, all pure-CPU, < 5s total:
  - PolicyNetwork shape sanity (logits + value + encoder_output) at `{64, 128, 256, 512}` × parametrize.
  - Bit-identity regression: `train_bc(hidden_size=64)` produces identical `eval_loss` / per-head accuracies / `joint_acc` on a tiny synthetic dataset across two calls with the same seed.
  - Monotone param-count growth across `{64, 128, 256, 512}` (guards the analyzer's `n_params` axis).

## Acceptance Criteria Verification

| Criterion | Status | Verification |
|---|---|---|
| Path A 2-cell `{64, 128}` driver | done | `run_issue280_hidden_size_sweep.sh` defaults to `HIDDEN_SIZES=(64 128)`, short budget |
| Path B accessible as fallback | done | `HIDDEN_SIZES="64 128 256 512" bash run_issue280_hidden_size_sweep.sh` |
| Analyzer keyed on `house_acc_on_work_subset` | done | `analyze_280.py:HOUSE_ON_WORK_CHANCE/SUCCESS`, `classify_verdict()` |
| Bit-identity test at default `hidden_size=64` | done | `test_train_bc_default_hidden_size_is_deterministic` |
| PolicyNetwork shape sanity at `{64, 128, 256, 512}` | done | `test_policy_network_accepts_hidden_size[*]` |
| Smoke test at `hidden_size=128` | done | Ran locally: 200 pairs / 2 epochs, n_params=23951, valid result.json |

## Test Plan

- All 6 unit tests pass locally: `python -m pytest tests/test_bc_fit_only_hidden_size.py -v`
- Verdict matrix verified by hand on synthetic rollups (capacity_not_gap, capacity_bound, information_gap, architecture_mismatch all classify correctly).
- bc_fit_only smoke at hidden_size=128 (50 env-steps × 2 epochs) runs cleanly and writes the expected JSON shape.
- Per user constraint, the sweep itself was NOT executed locally — it is intended for `COMPUTE_HOST_PRIMARY`.

## Notes for reviewer / executor

- This PR is built from `main`. PR #297 (#279, class-balanced BC) is still open. If #297 lands first and ships a `--class-balanced` flag, the executor can pass it through both the sweep driver and the verdict logic; the current PR is orthogonal to that axis.
- The `loom:blocked` label on #280 is informational (per user direction). Builder intentionally left it.

Closes #280

### CODE REVIEW

**Comment by rjwalters** [comment]
## Judge Evaluation: APPROVED

**Scope adherence (Path A)**: Curator scoped this to a 2-cell {64, 128} probe in light of PR #278's `gap_closed=0.934` at default `hidden_size=64`. PR respects scope:
- `run_issue280_hidden_size_sweep.sh:47` defaults to `HIDDEN_SIZES=(64 128)` with short budget (`NUM_STEPS=2500`, `EPOCHS=10` -> 10k pairs x 10 epochs, matching the #272/PR #276 failure regime).
- `analyze_280.py:236` defaults `--hidden-sizes` to `[64, 128]`.
- Path B (4-cell) accessible only via env override `HIDDEN_SIZES="64 128 256 512"`.

**No duplicate flag**: `bc_fit_only.py:287` already exposed `--hidden-size`; PR correctly leaves the script untouched and only adds driver + analyzer + tests.

**Tests (6/6 pass, 4.18s)**:
- Shape sanity at `{64, 128, 256, 512}` x 3 heads + value + encoder_output (`hidden_size`-shaped) - parametrized.
- Bit-identity at default `hidden_size=64` across two `train_bc` calls (NaN-aware on `house_acc_on_work_subset`).
- Monotone `n_params` growth across {64, 128, 256, 512} - guards analyzer's `n_params` axis.

**Verdict matrix** (`classify_verdict`): keyed on `house_acc_on_work_subset` with chance=0.354, success=0.90. Five buckets (`capacity_not_gap`, `information_gap`, `capacity_bound`, `architecture_mismatch`, `partial`) plus `no_data`. Order is sensible (saturation first -> chance-floor -> monotone-success -> ceiling -> mixed).

**Other quality signals**: git-rev stamping into result JSON for provenance; xargs `-P` parallelism matches #262 sweep idiom; runs in `tmux` on `COMPUTE_HOST_PRIMARY` per CLAUDE.md.

Ready for `loom:pr` -> human merge.
